src/IO/S3Common.cpp

```diff
@@ -107,89 +95,272 @@ class AWSLogger final : public Aws::Utils::Logging::LogSystemInterface
+    AWSEC2MetadataClient& operator =(const AWSEC2MetadataClient && rhs) = delete;
+    AWSEC2MetadataClient(const AWSEC2MetadataClient && rhs) = delete;
+
+    virtual ~AWSEC2MetadataClient() override
+    {
+    }
+
+    using Aws::Internal::AWSHttpResourceClient::GetResource;
+
+    virtual Aws::String GetResource(const char * resource_path) const
+    {
+        return GetResource(endpoint.c_str(), resource_path, nullptr/*authToken*/);
+    }
+
+    virtual Aws::String GetDefaultCredentials() const
     {
-        auto diff_seconds = static_cast<int>(Aws::Utils::DateTime::Now().SecondsWithMSPrecision() - Aws::Utils::DateTime(expiry.load()).SecondsWithMSPrecision());
-        if (diff_seconds > 0 - ACCOUNT_FOR_LATENCY)
+        std::unique_lock<std::recursive_mutex> locker(token_mutex);
+
+        LOG_TRACE(logger, "Getting default credentials for ec2 instance.");
+        auto result = GetResourceWithAWSWebServiceResult(endpoint.c_str(), EC2_SECURITY_CREDENTIALS_RESOURCE, nullptr);
+        Aws::String credentialsString = result.GetPayload();
+        auto httpResponseCode = result.GetResponseCode();
+
+        if (httpResponseCode == Aws::Http::HttpResponseCode::UNAUTHORIZED)
         {
-            auto * logger = &Poco::Logger::get("STSAssumeRoleCredentialsProvider");
-            LOG_DEBUG(logger, "Credentials have expired with diff of {} since last credentials pull.", diff_seconds);
-            LOG_TRACE(logger, "Grabbing lock.");
-            std::lock_guard<std::mutex> locker(credentials_mutex);
-            LOG_TRACE(logger, "Lock acquired. Checking expiration again.");
-            diff_seconds = static_cast<int>(Aws::Utils::DateTime::Now().SecondsWithMSPrecision() - Aws::Utils::DateTime(expiry.load()).SecondsWithMSPrecision());
-            if (diff_seconds > 0 - ACCOUNT_FOR_LATENCY)
-            {
-                auto caller_identity_outcome = sts_client->GetCallerIdentity({});
-                if (caller_identity_outcome.IsSuccess())
-                {
-                    String role_arn = caller_identity_outcome.GetResult().GetArn();
-                    if (role_arn != cached_role_arn || !underlying_credentials_provider)
-                    {
-                        underlying_credentials_provider = std::make_shared<Aws::Auth::STSAssumeRoleCredentialsProvider>(
-                            role_arn, session_name, String(), Aws::Auth::DEFAULT_CREDS_LOAD_FREQ_SECONDS, sts_client
-                        );
-                        cached_role_arn = role_arn;
-                    }
-                    cached_credentials = underlying_credentials_provider->GetAWSCredentials();
-                    expiry = Aws::Utils::DateTime::Now().SecondsWithMSPrecision();
-                }
-                else
-                {
-                    LOG_ERROR(logger, "Failed to make STS GetCallerIdentity request.");
-                }
-            }
+            return {};
+        }
+        locker.unlock();
```

Comment:
You don't have to call unlock. Use scopes.

Reply:
+